src/gui/accountstate.cpp

```diff
@@ -443,4 +400,62 @@ std::unique_ptr<QSettings> AccountState::settings()
     return s;
 }
 
+namespace {
+
+    class UpdateUrlDialog : public QMessageBox
+    {
+    public:
+        UpdateUrlDialog(AccountPtr account, const QUrl &newUrl, QWidget *parent)
+            : QMessageBox(QMessageBox::Warning,
+                tr("Url update requested for %1").arg(account->displayName()),
+                tr("The url for %1 changed from %2 to %3, do you want to accept the changed url?").arg(account->displayName(), account->url().toString(), newUrl.toString()),
+                QMessageBox::NoButton, parent)
+            , _account(account)
+            , _newUrl(newUrl)
+        {
+            setAttribute(Qt::WA_DeleteOnClose);
+            addButton(tr("Change url permanently to %1").arg(newUrl.toString()), QMessageBox::AcceptRole);
+            addButton(tr("Reject"), QMessageBox::RejectRole);
+            connect(this, &QMessageBox::accepted, this, &UpdateUrlDialog::acceptNewUrl);
+        }
+
+    public slots:
+        void acceptNewUrl()
+        {
+            _account->setUrl(_newUrl);
+            Q_EMIT _account->wantsAccountSaved(_account.data());
+        }
+
+    private:
+        AccountPtr _account;
+        const QUrl _newUrl;
+    };
+
+} // anonymous namespace
+
+void AccountState::updateUrlDialog(AccountPtr &account, const QUrl &newUrl)
+{
+    auto updateUrlDialog = new UpdateUrlDialog(account, newUrl, ocApp()->gui()->settingsDialog());
+
+    auto matchUrl = [](QUrl url1, QUrl url2) {
+        // ensure https://demo.owncloud.org/ matches https://demo.owncloud.org
+        // the empty path was the legacy formating before 2.9
+        if (url1.path().isEmpty()) {
+            url1.setPath(QStringLiteral("/"));
+        }
+        if (url2.path().isEmpty()) {
+            url2.setPath(QStringLiteral("/"));
+        }
+        return url1.matches(url2, QUrl::StripTrailingSlash | QUrl::NormalizePathSegments);
+    };
+
+    if (matchUrl(newUrl, account->url())) {
+        updateUrlDialog->accept();
+        delete updateUrlDialog;
```

Comment:
Maybe refactor acceptNewUrl to make the code accessible without without the need for a UpdateUrlDialog object.